packages/inter-protocol/src/vaultFactory/vault.js

```diff
@@ -480,77 +514,144 @@ export const prepareVault = (baggage, marshaller, zcf) => {
-          const giveMinted = AmountMath.min(
-            proposal.give.Minted || helper.emptyDebt(),
-            debt,
-          );
-          const wantMinted = proposal.want.Minted || helper.emptyDebt();
-          if (allEmpty([giveColl, giveMinted, wantColl, wantMinted])) {
-            clientSeat.exit();
-            return 'no transaction, as requested';
-          }
-
-          // Calculate the fee, the amount to mint and the resulting debt. We'll
-          // verify that the target debt doesn't violate the collateralization ratio,
-          // then mint, reallocate, and burn.
-          const { newDebt, fee, toMint } = helper.loanFee(
-            debt,
-            giveMinted,
-            wantMinted,
-          );
-
-          trace('adjustBalancesHook', state.idInManager, {
+          trace('adjustBalancesHook after quote', state.idInManager, {
             newCollateralPre,
-            newCollateral,
             fee,
             toMint,
+            maxDebtPre,
             newDebt,
           });
 
+          // While awaiting maxDebtFor, the allocations may have changed.
+          // After the `await`, we retrieve the vault's allocations again,
+          // so we can compare to the debt limit based on the new values.
+          const collateral = helper.getCollateralAllocated(vaultSeat);
+          const newCollateral = addSubtract(
+            collateral,
+            fp.give.Collateral,
+            fp.want.Collateral,
+          );
           if (
-            checkRestart(newCollateralPre, maxDebtPre, newCollateral, newDebt)
+            allocationsChangedSinceQuote(
+              newCollateralPre,
+              maxDebtPre,
+              newCollateral,
+              newDebt,
+            )
           ) {
             return helper.adjustBalancesHook(clientSeat);
           }
 
-          stageDelta(clientSeat, vaultSeat, giveColl, wantColl, 'Collateral');
+          return helper.commitBalanceAdjustment(
+            clientSeat,
+            fp,
+            {
+              newDebt,
+              fee,
+              surplus,
+              toMint,
+            },
+            { normalizedDebtPre, collateralPre },
+          );
+        },
+
+        /**
+         *
+         * @param {ZCFSeat} clientSeat
+         * @param {FullProposal} fp
+         * @param {ReturnType<typeof calculateLoanCosts>} costs
+         * @param {object} accounting
+         * @param {NormalizedDebt} accounting.normalizedDebtPre
+         * @param {Amount<'nat'>} accounting.collateralPre
+         * @returns {Promise<string>} success message
+         */
+        async commitBalanceAdjustment(
+          clientSeat,
+          fp,
+          { newDebt, fee, surplus, toMint },
+          { normalizedDebtPre, collateralPre },
+        ) {
+          const { state, facets } = this;
+          const { helper } = facets;
+          const { vaultSeat } = state;
+
+          const takeMinted = AmountMath.subtract(fp.give.Minted, surplus);
```

Comment:
**take**Minted is counter to all my instincts. This needs a comment to explain where it fits with `give` and `want`.   It looks what I might have call `actualGiveMinted`. Is it that or something else?

Reply:
My thought was "we're not taking all the the offer is giving, we'll just take this much Minted". In other places that the contract takes less than is offered, how is it described? "actualGive" has a logic but I read "give" always as specified by the offer. I think that's good to maintain in the coding style. "Take" has a logic: it's what the contract is doing with the give.

`giveMintedTaken` ? `takeOfGive`? Or is a comment sufficient?

Reply:
A comment would be sufficient. I'm ok with `giveMintedTaken`